### PR TITLE
docs: Add some guidelines for structs/classes/variables

### PR DESCRIPTION
**Pull request by danobi**
These help with maintaining some uniformity in code. I think the codebase is already pretty good about these rules, so it's just codifying it.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

### CODE REVIEW

File: docs/coding_guidelines.md [hunk 1]
```diff
@@ -49,3 +49,20 @@ the kernel debug facilities. And we probably should not be mounting
...
+Use a struct only for passive objects that carry data; everything else is a class.
+
+All fields in a struct should be public. The struct must not have invariants
+that imply relationships between different fields, since direct user access to
+those fields may break those invariants. Structs should not have any associated
+methods (including constructors and destructors).
+
+## Variable naming
+
+Variables should be in `snake_case` (all lowercase, with spaces between words).
```
Comment (jordalgo) [inline docs/coding_guidelines.md:64]:
```suggestion
Variables should be in `snake_case` (all lowercase, with underscores between words).
```

File: docs/coding_guidelines.md [hunk 2]
```diff
@@ -49,3 +49,20 @@ the kernel debug facilities. And we probably should not be mounting
...
+
+Class data members should have a trailing underscore (for both public and private).
```
Comment (viktormalik) [inline docs/coding_guidelines.md:66]: I don't want to start a dispute over coding conventions but what is the motivation for this? I always thought that the trailing underscore is to visually distinguish private members from the public ones.

Comment (jordalgo) [reply docs/coding_guidelines.md:66]: TBH, I thought this too but then couldn't think of very many instances where we have a public class data members as the general c++ best practice is to hide them behind getter/setters.

Comment (viktormalik) [reply docs/coding_guidelines.md:66]: We have quite a few in the `BPFtrace` class: `resources`, `maps`, `structs`, `macros_`, `enums_` (they don't even follow any rules). Same for all members of `RequiredResources` (where it doesn't really make sense to hide them behind getters/setters).

Comment (jordalgo) [reply docs/coding_guidelines.md:66]: Sorry, my thought was completely muddled. I was trying to agree with you @viktormalik but also mention that perhaps we should abide by some best practices and try to hide these members behind getters/setters or at least abide by some general code guideline around when to make class variables public (e.g. if they are read-only they should always be const -- stuff like that).

Comment (danobi) [reply docs/coding_guidelines.md:66]: At a high level I'd like to point out this is a purely style thing. I don't recall us ever having any issues with breaking invariants. 

The underscore part was copied from google's C++ coding guidelines. If we find trailing `_` more useful for public/private distinction, I think we should be free to do that. In fact, I'm thinking that it's more useful in general to do that.

I like the idea of getter/setters but, yeah, for RequiredResources it seems like it'd just be a lot of noise.

Comment (ajor) [reply docs/coding_guidelines.md:66]: I like the idea of only using a trailing underscore for private members and none for public members.

This got me digging into the Google C++ Style Guide, and if I'm understanding correctly, while they do require underscores for all members, they also require that all data members be private:
https://google.github.io/styleguide/cppguide.html#Variable_Names
https://google.github.io/styleguide/cppguide.html#Access_Control

Don't think we need to go as far as requiring no public members though.

**Comment by danobi** [comment]
Ready for another loo

**Review by viktormalik — APPROVED** [review]
Nice and clear!

If only the existing code followed these guidelines... :thinking:

**Comment by jordalgo** [comment]
LGTM!

**Review by jordalgo — APPROVED** [review]

**Pull request merged**